src/Oscoin/Protocol/Sync.hs

```diff
@@ -354,109 +386,317 @@ getRemoteHeight
     => SyncT c tx s m Height
 getRemoteHeight = map height getRemoteTip
 
--- | Given a valid 'Range', syncs the blocks within that range.
+{------------------------------------------------------------------------------
+                            Syncing blocks
+
+The syncing process starts by calling 'commonChainHeight' to retrieve the
+height associated to the block which root hash matches across all peers,
+including the local node. To understand why this is important, let's take a
+look at this picture, which shows 3 nodes at different heights and on different
+chains:
+
+
+                    height
+                               +-----+
+              node1   0 -------|-----|----------------------- 100
+                               |     |
+                               |     |
+                               |     |
+                               |     |
+                               |     |     70
+              node2   0 -------|-----|------\
+                               |     |       \
+                               |     |        \
+                               |     |         --------- 80
+                               |     |
+                               |    -|-- 60
+                               |   / |
+              node3   0 -------|--/  |
+                               | 55  |
+                               +-----+
+
+By returning 55 (perhaps inclusive of the block and/or the hash associated
+with it) we are sure we are starting from a valid block.
+
+-}
+
+-- | Calculates the \"common chain height\" between this node and its active
+-- peers. In case consensus can't be reached, the function retries again with
+-- a fibonacci sequence, i.e. by requesting @local_tip - fib(n)@ up to a
+-- limit.
+-- Returns 0 (i.e. genesis' height) if no common height could be retrieved.
+commonChainHeight
+    :: forall c tx s m.
+       ( Monad m
+       , Ord (Hash c)
+       )
+    => Block c tx (Sealed c s)
+    -> SyncT c tx s m Height
+commonChainHeight = commonChainHeightWith defaultBacktrackPolicy
+  where
+      defaultBacktrackPolicy :: BacktrackPolicy
+      defaultBacktrackPolicy = fib
+
+type BacktrackPolicy = Int -> Int
+
+linearBacktrackPolicy :: BacktrackPolicy
+linearBacktrackPolicy = succ
+
+-- | Allows the 'commonChainHeight' algorithm to be passed a 'RetryPolicy'
+-- which determines how far we need to backtrack when calculating the new
+-- 'Height'.
+commonChainHeightWith
+    :: forall c tx s m.
+       ( Monad m
+       , Ord (Hash c)
+       )
+    => BacktrackPolicy
+    -> Block c tx (Sealed c s)
+    -> SyncT c tx s m Height
+commonChainHeightWith policy localTip = do
+    ctx <- ask
+    go ctx (height localTip) 0 5
```

Comment:
IIUC, you're not using the `retry` package here due to `threadDelay` being mockable. I still wonder if this could be made a bit clearer - and allow for instrumentation - if the recursion could be factored out of the actual work. As in `go :: SyncContext c tx s m -> Height -> SyncT c tx s m (Either RetryCondition Height)` + a `loop` function which scrutinises `RetryCondition` and either emits some telemetry and retries, or returns `Height`.

Reply:
> IIUC, you're not using the retry package here due to threadDelay being mockable.

Yes, that's exactly it. Unfortunately all the combinators from `retry` lives in `MonadIO`, which kills any hope to use the package in a pure setting. I did also try briefly to use _only_ the `RetryStatus` and `RetryPolicy` parts (which are pure) and integrate them into our code, but the result was horrible, as those have a very specific use case, tailored to that library.

> I still wonder if this could be made a bit clearer - and allow for instrumentation - if the recursion could be factored out of the actual work. As in go :: SyncContext c tx s m -> Height -> SyncT c tx s m (Either RetryCondition Height) + a loop function which scrutinises RetryCondition and either emits some telemetry and retries, or returns Height.

Maybe -- I will try rehashing things and see how things turns out.